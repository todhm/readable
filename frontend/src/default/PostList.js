import React, {Fragment, Component} from 'react'
import {Button, Card, ListItemText} from '@material-ui/core';
import {ArrowDownward, ArrowUpward} from '@material-ui/icons';
import {ThumbUp, ThumbDown} from '@material-ui/icons';
import {Link} from 'react-router-dom';
import LittlePost from './LittlePost';

export default class extends Component {

    state = {
        desc: false
    }

    handleArrow = (e) => {
        this.setState(({desc}) => ({
            desc: !desc
        }))
    }

    render() {
        const {desc} = this.state;
        const {postList, sortMethod} = this.props;
        const sortedPostList = sortMethod
            ? postList.sort(
                (a, b) => desc
                ? b[sortMethod] - a[sortMethod]
                : a[sortMethod] - b[sortMethod])
            : postList

        return (<Fragment>
            {
                sortMethod
                    ? [
                        desc
                            ? <Button onClick={this.handleArrow}>
                                <ArrowUpward/>
                            </Button>
                            : <Button onClick={this.handleArrow}>
                                <ArrowDownward/>
                            </Button>
                    ]
                    : null
            }

            <Card>
                {
                    sortedPostList.map(post => (
                        typeof post.title === 'string' && !post.deleted
                        ? <LittlePost category={post.category} title={post.title} body={post.body} author={post.author} postId={post.id} key={post.id} voteScore={post.voteScore} voteUp={post.votedUp} voteDown={post.votedDown} commentList={post.commentList}/>
                        : null))
                }
            </Card>
        </Fragment>)
    }
}
