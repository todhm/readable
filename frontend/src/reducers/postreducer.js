import * as postAction from '../post/actionTypes'

export const initialPost = {
	id: "",
	timestamp: 0,
	title: "",
	body: "",
	author: "",
	category: "",
	voteScore: 1,
	deleted: false,
	votedUp: false,
	votedDown: false,
	commentList: [],
}

const  PostReducer = (state = initialPost, action) => {
	switch (action.type) {
		default: return state
		case postAction.GET_POST:
				return {
				...state,
				...action.data
			}

		case postAction.GET_COMMENTS:
				return {
				...state,
				commentList: action.data
			}

		case postAction.ADD_COMMENTS:
				return {
				...state,
				commentList: [
					...state.commentList,
					action.data
				]
			}
		case postAction.EDIT_COMMENT:
				return {
				...state,
				commentList: [
					...state.commentList.map((comment, index) => {
						if (comment.id !== action.commentId) {
							return comment;
						}
						return {
							...action.data
						};
					})
				]
			}

		case postAction.DELETE_COMMENT:
				return {
				...state,
				commentList: [
					...state.commentList.filter((comment) =>
						comment.id != action.commentId
					)
				]
			}
	}
}

export default PostReducer;
