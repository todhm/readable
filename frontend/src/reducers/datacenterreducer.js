import * as defaultAction from '../default/actionTypes'


const dataCenter = {
	categoryList: {},
	postList: {},

}


const DataCenterReducer = (state = dataCenter, action) => {
	switch (action.type) {
		default: return state

		case defaultAction.GET_CATEGORIES:
				return {
				...state,
				categoryList: action.categoryList,
			}
		case defaultAction.GET_POSTS:
				return {
				...state,
				postList: action.postList,
			}


		case defaultAction.ADD_POST:
				return {
				...state,
				postList: {
					...state.postList,
					[action.payload.id]: {
						...action.payload
					}
				},
			}

		case defaultAction.GET_COMMENTS:
				return {
				...state,
				postList: {
					...state.postList,
					[action.postId]: {
						...state.postList[action.postId],
						commentList: action.payload,
					}
				}
			}

		case defaultAction.UPDATE_POST:
				return {
				...state,
				postList: {
					...state.postList,
					[action.payload.id]: {
						...action.payload
					}
				},
			}

		case defaultAction.DELETE_POST:
				return Object.assign({}, state, {
				postList: Object.keys(state.postList).reduce((result, id) => id === action.postId ?
					result : { ...result,
						[id]: state.postList[id]
					}, {})
			})


		case defaultAction.UPDATE_CATEGORY_SORT:
				return {
				...state,
				categoryList: {
					...state.categoryList,
					[action.category]: {
						...state.categoryList[action.category],
						sortMethod: action.sortMethod
					}
				}
			}
	}

}

export default DataCenterReducer;
