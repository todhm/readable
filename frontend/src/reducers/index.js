import {combineReducers} from 'redux'
import * as defaultAction from '../default/action'
import * as postAction from '../post/action'
const dataCenter={
    categoryList:{},
    postList:{},

}

export const initialPost ={
    id:"",
    timestamp:0,
    title:"",
    body:"",
    author:"",
    category:"",
    voteScore:1,
    deleted:false,
    votedUp:false,
    votedDown:false,
    commentList:[],
}

export const PostReducer =(state=initialPost,action)=>{
    switch(action.type){
        default:
            return state
        case postAction.GET_POST:
            return {
                ...state,
                ...action.data
            }

        case postAction.GET_COMMENTS:
            return {
                ...state,
                commentList:action.data
            }

        case postAction.ADD_COMMENTS:
            return {
                ...state,
                commentList:[
                    ...state.commentList,
                    action.data
                ]
            }
        case postAction.EDIT_COMMENT:
            return {
                ...state,
                commentList:[
                    ...state.commentList.map((comment,index)=>{
                        if(comment.id !== action.commentId) {
                            return comment;
                        }
                        return {
                            ...action.data
                        };
                    })
                ]
            }

        case postAction.DELETE_COMMENT:
            return {
                ...state,
                commentList:[
                    ...state.commentList.filter((comment)=>
                        comment.id!=action.commentId
                    )
                ]
            }
    }
}

const CommentReducer =(state=initialComment,action)=>{
    switch(action.type){
        default:
            return state

    }
}

export const DataCenterReducer =(state=dataCenter,action)=>{
    switch(action.type){
        default:
            return state

        case defaultAction.GET_CATEGORIES:
            return {
                ...state,
                categoryList:action.categoryList,
            }
        case defaultAction.GET_POSTS:
            return {
                ...state,
                postList:action.postList,
            }


        case defaultAction.ADD_POST:
            return {
                ...state,
                postList:{
                    ...state.postList,
                    [action.payload.id]:{
                        ...action.payload
                    }
                },
            }

        case defaultAction.GET_COMMENTS:
            return {
                ...state,
                postList:{
                    ...state.postList,
                    [action.postId]:{
                        ...state.postList[action.postId],
                        commentList:action.payload,
                    }
                }
            }


        case defaultAction.UPDATE_CATEGORY_SORT:
            return {
                ...state,
                categoryList:{
                    ...state.categoryList,
                    [action.category]:{
                        ...state.categoryList[action.category],
                        sortMethod:action.sortMethod
                    }
                }
            }
    }

}


export default combineReducers({
    DataCenterReducer,
    PostReducer
});
