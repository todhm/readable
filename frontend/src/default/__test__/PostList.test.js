import React from 'react';
import renderer from 'react-test-renderer';
import PostList from '../PostList';
import {initialPost} from '../../reducers'
import { shallow } from 'enzyme';

const postList=[initialPost]
const sortMethod=""
const mockedEvent = {
      target: {
        value: 'world'
      }
    };
describe('PostList', () => {
  let component = null;

  it('no problem rendering', () => {

    component = shallow(
        <PostList postList={postList} sortMethod={"timestamp"} />
    );
  });

  it('same with snapshot ', () => {
    expect(component).toMatchSnapshot();
  })


  it('handle arrow correctly', () => {
      const instance  = component.instance()
      instance.handleArrow(mockedEvent);
      expect(instance.state.desc).toBe(true); // value 값이 2인지 확인
  });

});
