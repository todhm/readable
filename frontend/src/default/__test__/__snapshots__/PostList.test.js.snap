// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList same with snapshot  1`] = `
<React.Fragment>
  <WithStyles(Button)
    onClick={[Function]}
  >
    <pure(ArrowDownward) />
  </WithStyles(Button)>
  <WithStyles(Card)>
    <WithStyles(Connect(LittlePost))
      postId=""
      title=""
      voteDown={false}
      voteScore={1}
      voteUp={false}
    />
  </WithStyles(Card)>
</React.Fragment>
`;
